ops/lib/__init__.py

```diff
@@ -108,53 +113,92 @@ def _find_all_specs(path):
             try:
                 lib_dirs = os.listdir(opslib)
             except OSError:
+                # logger.trace(...) # *lots* of things checked here
                 continue
+            else:
+                logger.debug("Trying %r.", opslib)
             finder = get_importer(opslib)
-            if finder is None or not hasattr(finder, 'find_spec'):
+            if finder is None:
+                logger.debug("Finder for %r is None.", opslib)
+                continue
+            if not hasattr(finder, 'find_spec'):
+                logger.debug("Finder for %r has no find_psec.", opslib)
                 continue
             for lib_dir in lib_dirs:
-                spec = finder.find_spec("{}.opslib.{}".format(top_dir, lib_dir))
+                spec_name = "{}.opslib.{}".format(top_dir, lib_dir)
+                spec = finder.find_spec(spec_name)
                 if spec is None:
+                    logger.debug("No spec for %r.", spec_name)
                     continue
                 if spec.loader is None:
                     # a namespace package; not supported
+                    logger.debug("No loader for %r (probably a namespace package)", spec_name)
                     continue
+
+                logger.debug("Found %r.", spec_name)
                 yield spec
 
 
 # only the first this many lines of a file are looked at for the LIB* constants
 _MAX_LIB_LINES = 99
 
 
+_EXPECTED = {'NAME': str, 'AUTHOR': str, 'API': int, 'PATCH': int}
+
+
+# XXX Names Are Hard™
+class _NAH:
+    """A silly little helper to only work out the difference between
+    what was found and what was expected when logging"""
+
+    def __init__(self, found):
+        self._found = found
+
+    def __str__(self):
+        return ", ".join(set(_EXPECTED) - set(self._found))
+
+
 def _parse_lib(spec):
     if spec.origin is None:
+        # "can't happen"
+        logger.debug("No origin for %r (no idea why; please report)", spec.name)
         return None
 
-    _expected = {'NAME': str, 'AUTHOR': str, 'API': int, 'PATCH': int}
+    logger.debug("Parsing %r:", spec.name)
 
     try:
         with open(spec.origin, 'rt', encoding='utf-8') as f:
             libinfo = {}
             for n, line in enumerate(f):
-                if len(libinfo) == len(_expected):
+                if len(libinfo) == len(_EXPECTED):
                     break
                 if n > _MAX_LIB_LINES:
+                    logger.debug(
+                        "  Reached line %d without finding %s.",
```

Comment:
I like this sort of indenting, to attach it to the rest.
I wonder about:
```
found %s, didn't find %s
```

Thoughts?

Reply:
good idea, i'll tweak the helper class __str__ to produce that